src/utilities.jl

```diff
@@ -16,10 +16,22 @@ function Base.randn(rng::AbstractVector{<:AbstractRNG}, T, dim::Int, n_chains::I
     return cat(randn.(rng, T, dim)...; dims=2)
 end
 
-# Sample from Categorical distributions
+# Helper functions to sync RNGs that ensures coupling works properly and reproducible
 
-randcat_logp(rng::AbstractRNG, unnorm_ℓp::AbstractVector) =
-    randcat(rng, exp.(unnorm_ℓp .- logsumexp(unnorm_ℓp)))
+rand_sync(rng::AbstractRNG, args...) = rand(rng, args...)
+
+# This function is needed to use only one RNG while a vector of RNGs are provided
+# in a way that RNGs are still synchronised. This is important if we want to use 
+# same RNGs to couple multiple chains.
+function rand_sync(rngs::AbstractVector{<:AbstractRNG}, args...)
+    # Dummpy calles to sync RNGs
+    foreach(rngs[2:end]) do rng
```

Comment:
This is potentially a waste of computation. Consider to simply "duplicate" the first RNG so that all RNGS are identical. For example:

```
rngs[2:end] = [copy(rngs[1]) for i=1:(length(rngs)-1)]) 
```

Reply:
What we have is a slightly different issue. Let me explain if from the begining: 

Assume we have two chains with `rng1` and `rng2`. When we are using coupled HMC, what is pass is `rng = [rng1, rng2]` where `rng1 == rng2`. Is this all good for *two* rand call for *two* chains but not to do *one* rand call for *two* chains.

To make this intuitive, consider below are computations for two chains

```
chain 1: ... -> rand(rng1) -> ... -> rand(rng1)
chain 2: ... -> rand(rng2) -> ... -> rand(rng2)
```
is a situation works fine.

But 
```
chain 1: ... -> rand(rng1) -\                  /-> rand(rng1)
                                              |> rand(???) -
chain 2: ... -> rand(rng2) -/                  \-> rand(rng2)
```
is a situation we were not --- we simply using `GLOBAL_RNG` for `???` before, which is correct but doesn't make the computation reproducible. What I'm trying to fix is this reproducibiltiy.

Come back to what you propose, yes, it will fix the issue *if* we do want to use the *identical* rngs for all chains. However, if we did so, our vectorized implementation will enforce all chains to be coupled there, which is not what we want: what we are is, if the passed-in rngs are idential, make sure the rng sequences are same for two chans; if the passed-in rngs are not idential, we don't want to introduce any coupling.

Ideally, we need a coupled chain level rng to do so, i.e. let `??? = rng_coupled` or something. But I don't think this is easy to be done with the current abstraction. Also I think this computation waste is negectble.

Reply:
`copy` an RNG will produce an independent copy of that RNG, so copying leads to identical but de-coupled random number generators. That said, I'm happy to merge this as-is for now.